Initialize directory structure

window9u opened the pull request:
### **What this PR does / Why we need it**  
This PR defines the directory structure and clarifies the role of each component in the project.  

### **Issue(s) Fixed**  
Fixes #430  

### **Special Notes for Reviewers**  
- Added a detailed `design/directory_structure.md` document.  
- Standardized module responsibilities in `internal/` directory.  

### **User-Facing Changes?**  
```release-note
NONE
```  

### **Directory Structure**  

This PR establishes the following directory structure:  

```
project-root/
├── api        # API documentation & DTO definitions
├── bin        # Compiled binaries
├── cmd        # Application entry points
│   └── codepair  # Main application entry point (main.go)
├── design     # Design documents and architecture references
└── internal   # Core application logic (business logic, infrastructure, middleware)
```

#### **Internal Structure Breakdown**  
```
internal/
├── config        # Configuration loading & validation
├── domain        # Business logic (auth, user, etc.)
├── infra         # Infrastructure dependencies (DB, storage, external services)
├── middleware    # Shared middleware (logging, auth, validation)
└── server        # Server setup, routing, and bootstrapping
```

### **Key Components**  

#### 1. `internal/config`  
- Loads and validates configurations (e.g., `viper`, `os.Getenv`).  

#### 2. `internal/domain`  
- Business logic and domain entities (`auth`, `user`).  
- Structure:  
  - `model.go`: Entity definitions  
  - `repository.go`: Database interactions  
  - `service.go`: Business logic  
  - `handler.go`: HTTP handlers  

#### 3. `internal/infra`  
- External dependencies (DB, storage, external services).  
- Examples:  
  - `database/mongo/`: MongoDB repository implementations  
  - `storage/`: MinIO, S3 integrations  
  - `yorkie/`: Yorkie API client for collaborative editing  

#### 4. `internal/middleware`  
- Common middleware (logging, auth, error handling).  

#### 5. `internal/server`  
- Server setup and routing (`routes.go`, `server.go`).  

---

### **Checklist**  
- [x] Added relevant tests or not required  
- [x] No breaking changes  



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced a structured server configuration system.
  - Added foundational server initialization process.
  - Established project directory structure documentation.

- **Infrastructure**
  - Created initial project layout with dedicated directories for configuration, core logic, and server management.
  - Prepared MongoDB database infrastructure placeholder.

- **Configuration**
  - Implemented default server port configuration (Port: 3001).
  - Added configuration loading mechanism.

- **Server**
  - Set up basic routing with a "Hello, World!" endpoint.
  - Implemented server startup and error handling.

- **User Management**
  - Initialized user management package structure.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

window9u commented on backend-go/design/directory_structure.md:
````diff
@@ -0,0 +1,75 @@
+
+```
+backend-go/
+├── api        # API documentation & DTO definitions
````
I'm not sure `OpenAPI` document would be here.
If we also generate frontend client code by `OpenAPI` docs, I think it could be placed in root-directory

kokodak replied:
I also think it would be better to place the OpenAPI document in the root directory.
Having a single OpenAPI document to manage both the frontend and backend code provides a clear context, making it more appropriate to keep it at the root level.

devleejb replied:
I agree~

devleejb commented on backend-go/internal/server/server.go:
```diff
@@ -0,0 +1,63 @@
+
+	RegisterRoutes(e)
+
+	s := &http.Server{
+		Addr:        fmt.Sprintf(":%d", conf.Server.Port),
+		Handler:     e,
+		ReadTimeout: conf.Server.ReadTimeout,
+	}
```
Q) Why should we use http.Server? Can’t we just use `echo` instead?

devleejb commented on backend-go/internal/server/server.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+func (c *Codepair) Start() error {
+	errCh := make(chan error, 1)
+	go func() {
+		log.Printf("Server starts running on %d\n", c.config.Server.Port)
+		if err := c.httpServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+			errCh <- fmt.Errorf("server error: %v", err)
+		}
+	}()
+	select {
+	case err := <-errCh:
+		return err
+	case <-time.After(100 * time.Millisecond):
+		return nil
+	}
```
I think we can just call `e.Start(Port)`.

devleejb commented on backend-go/internal/config/server.go:
```diff
@@ -0,0 +1,13 @@
+
+const (
+	DefaultServerPort  = 3001
+	DefaultReadTimeout = 10 * time.Second
```
I think it need to be increased due to Yorkie Intelligence.

devleejb commented on backend-go/cmd/codepair/main.go:
```diff
@@ -1,17 +1,34 @@
-	e.Logger.Fatal(e.Start(":3001"))
+	conf := config.LoadConfig()
+
+	cp := server.New(conf)
+
+	if err := cp.Start(); err != nil {
+		log.Printf("server error: %v", err)
+		os.Exit(1)
+	}
+
+	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer stop()
+	<-ctx.Done()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	if err := cp.Shutdown(ctx); err != nil {
+		log.Printf("server shutdown error: %v", err)
+	}
```
I think we can reduce these code if we use `echo` server.

devleejb reviewed (changes requested):
Thank you for your contribution! I left a few comments.

window9u commented:
**@devleejb** Thank you for your review! I’ve applied your recommendations:  

- Removed `readTimeout` (I believe it should be handled when setting up middlewares).  
- Replaced `net/http` with Echo.  
- Set the graceful shutdown duration to **30 seconds**.  

Initially, I used `net/http` for its configurability. However, after further research, I found that Echo provides most of the same functionalities, such as [timeout middleware](https://echo.labstack.com/docs/middleware/timeout#usage) and [graceful shutdown](https://echo.labstack.com/docs/cookbook/graceful-shutdown#using-httpservershutdown).  

Thanks again for your valuable feedback! 🚀

devleejb commented on backend-go/cmd/codepair/main.go:
```diff
@@ -1,17 +1,18 @@
+	cp := server.New(conf)
+
+	if err := cp.Start(); err != nil {
+		log.Fatalf("Failed to start server: %v", err)
```
I think the following redundant error message might appear:
```
Failed to start server: failed to start server: ...
```
How about simplifying it with `log.Fatal(err)`?

window9u replied:
I applied that. Thank you for check my mistake!

devleejb reviewed (approved):
LGTM
